apps/aecore/lib/aecore/contract/contract_state_tree.ex

```diff
@@ -0,0 +1,113 @@
+defmodule Aecore.Contract.ContractStateTree do
+  @moduledoc """
+  Top level contract state tree.
+  """
+  alias Aecore.Chain.Identifier
+  alias Aecore.Contract.Contract
+  alias Aeutil.PatriciaMerkleTree
+  alias Aeutil.Serialization
+
+  @contract_key_size 32
+
+  @type hash :: binary()
+  @type contracts_state() :: Trie.t()
+
+  @spec init_empty() :: contracts_state()
+  def init_empty do
+    PatriciaMerkleTree.new(:contracts)
+  end
+
+  @spec insert_contract(contracts_state(), Contract.t()) :: contracts_state()
+  def insert_contract(contract_tree, %Contract{id: id, store: store} = contract) do
+    serialized = Serialization.rlp_encode(contract)
+
+    new_contract_tree = PatriciaMerkleTree.insert(contract_tree, id.value, serialized)
+    store_id = Contract.store_id(contract)
+
+    Enum.reduce(store, new_contract_tree, fn {s_key, s_value}, tree_acc ->
+      s_tree_key = <<store_id::binary, s_key::binary>>
+      PatriciaMerkleTree.insert(tree_acc, s_tree_key, s_value)
+    end)
+  end
+
+  @spec enter_contract(contracts_state(), Contract.t()) :: contracts_state()
+  def enter_contract(contract_tree, %Contract{id: id, store: store} = contract) do
+    serialized = Serialization.rlp_encode(contract)
+
+    updated_contract_tree = PatriciaMerkleTree.enter(contract_tree, id.value, serialized)
+    store_id = Contract.store_id(contract)
+    old_contract_store = get_store(store_id, contract_tree)
+
+    update_store(store_id, old_contract_store, store, updated_contract_tree)
+  end
+
+  @spec get_contract(contracts_state(), binary()) :: Contract.t()
+  def get_contract(contract_tree, key) do
+    case PatriciaMerkleTree.lookup(contract_tree, key) do
+      {:ok, serialized} ->
+        {:ok, deserialized} = Serialization.rlp_decode_anything(serialized)
+
+        identified_id = Identifier.create_identity(key, :contract)
+
+        identified_referers =
+          Enum.map(deserialized.referers, fn referer ->
+            {:ok, identified_referer} = Identifier.create_identity(referer, :contract)
```

Comment:
This seams weird. Don't we store encoded Identifier inside contract encoding and decode it? I think after decoding we already have an Identifier structure in decoded contract.

Reply:
Yes, my mistake. I tested and fixed it just now.